Create dotnet-dsrouter.md (#27288)
dotnet/docs: docs/core/diagnostics/dotnet-dsrouter.md

## Options

- **`-?|-h|--help`**

  Shows command-line help.

- **`--version`**

  Displays the version of the `dotnet-dsrouter` utility.

## Commands

| Command                                                         |
|-----------------------------------------------------------------|
| [dotnet-dsrouter client-server](#dotnet-dsrouter-client-server) |
| [dotnet-dsrouter server-server](#dotnet-dsrouter-server-server) |
| [dotnet-dsrouter server-client](#dotnet-dsrouter-server-client) |
| [dotnet-dsrouter client-client](#dotnet-dsrouter-client-client) |

## dotnet-dsrouter client-server

Start a .NET application Diagnostics Server routing local IPC server <--> remote TCP client. Router is configured using an IPC client (connecting diagnostic tool IPC server) and a TCP/IP server (accepting runtime TCP client).

### Synopsis

```console
dotnet-dsrouter client-server
    [-ipcc|--ipc-client <ipcClient>]
    [-tcps|--tcp-server <tcpServer>]
    [-rt|--runtime-timeout <timeout>]
    [-v|--verbose <level>]
    [-fp|--forward-port <platform>]
```

### Options

- **`-ipcc, --ipc-client <ipcClient>`**

>   The diagnostic tool diagnostics server ipc address (--diagnostic-port argument). Router connects diagnostic tool ipc server when establishing a new route between runtime and diagnostic tool.

Reviewer: ```suggestion
  The diagnostic tool diagnostics server IPC address (--diagnostic-port argument). Router connects diagnostic tool IPC server when establishing a new route between runtime and diagnostic tool.
```

- **`-tcps, --tcp-server <tcpServer>`**

>   The router TCP/IP address using format [host]:[port]. Router can bind one (127.0.0.1, [::1], 0.0.0.0, [::], ipv4 address, ipv6 address, hostname) or all (*) interfaces. Launch runtime using `DOTNET_DiagnosticPorts` environment variable connecting router TCP server during startup.

Reviewer: ```suggestion
  The router TCP/IP address using format `[host]:[port]`. Router can bind one (`127.0.0.1`, `[::1]`, `0.0.0.0`, `[::]`, IPv4 address, IPv6 address, hostname) or all (*) interfaces. Launch runtime using `DOTNET_DiagnosticPorts` environment variable connecting router TCP server during startup.
```

- **`-rt, --runtime-timeout <runtimeTimeout>`**
  Automatically shut down router if no runtime connects to it before specified timeout (seconds). If not specified, router won't trigger an automatic shutdown.

- **`-v, --verbose <verbose>`**
  Enable verbose logging (debug|trace)

- **`-fp, --forward-port <forwardPort>`**
  Enable port forwarding, values Android|iOS for TcpClient and only Android for TcpServer. Make sure to set `ANDROID_SDK_ROOT` before using this option on Android.

## dotnet-dsrouter server-server

Start a .NET application Diagnostics Server routing local IPC client <--> remote TCP client. Router is configured using an IPC server (connecting to by diagnostic tools) and a TCP/IP server (accepting runtime TCP client).

### Synopsis

```console
dotnet-dsrouter server-server
    [-ipcs|--ipc-server <ipcServer>]
    [-tcps|--tcp-server <tcpServer>]
    [-rt|--runtime-timeout <timeout>]
    [-v|--verbose <level>]
    [-fp|--forward-port <platform>]
```

### Options

- **`-ipcs, --ipc-server <ipcServer>`**

>   The diagnostics server ipc address to route. Router accepts ipc connections from diagnostic tools establishing a new route between runtime and diagnostic tool. If not specified router will use default ipc diagnostics server path.

Reviewer: ```suggestion
  The diagnostics server IPC address to route. Router accepts IPC connections from diagnostic tools establishing a new route between runtime and diagnostic tool. If not specified router will use default IPC diagnostics server path.
```

- **`-tcps, --tcp-server <tcpServer>`**

>   The router TCP/IP address using format [host]:[port]. Router can bind one (127.0.0.1, [::1], 0.0.0.0, [::], ipv4 address, ipv6 address, hostname) or all (*) interfaces. Launch runtime using `DOTNET_DiagnosticPorts` environment variable connecting router TCP server during startup.

Reviewer: ```suggestion
  The router TCP/IP address using format `[host]:[port]`. Router can bind one (`127.0.0.1`, `[::1]`, `0.0.0.0`, `[::]`, IPv4 address, IPv6 address, hostname) or all (*) interfaces. Launch runtime using `DOTNET_DiagnosticPorts` environment variable connecting router TCP server during startup.
```

- **`-v, --verbose <verbose>`**
  Enable verbose logging (debug|trace)

- **`-fp, --forward-port <forwardPort>`**
  Enable port forwarding, values Android|iOS for TcpClient and only Android for TcpServer. Make sure to set `ANDROID_SDK_ROOT` before using this option on Android.

## dotnet-dsrouter server-client

Start a .NET application Diagnostics Server routing local IPC client <--> remote TCP server. Router is configured using an IPC server (connecting to by diagnostic tools) and a TCP/IP client (connecting runtime TCP server).

### Synopsis

```console
dotnet-dsrouter server-client
    [-ipcs|--ipc-server <ipcServer>]
    [-tcpc|--tcp-client <tcpClient>]
    [-rt|--runtime-timeout <timeout>]
    [-v|--verbose <level>]
    [-fp|--forward-port <platform>]
```

### Options

- **`-ipcs, --ipc-server <ipcServer>`**
  The diagnostics server ipc address to route. Router accepts ipc connections from diagnostic tools establishing a new route between runtime and diagnostic tool. If not specified router will use default ipc diagnostics server path.

- **`-tcpc, --tcp-client <tcpClient>`**

>   The runtime TCP/IP address using format [host]:[port]. Router can connect 127.0.0.1, [::1], ipv4 address, ipv6 address, hostname addresses. Launch runtime using `DOTNET_DiagnosticPorts` environment variable to setup listener

Reviewer: ```suggestion
  The runtime TCP/IP address using format `[host]:[port]`. Router can connect `127.0.0.1`, `[::1]`, IPv4 address, IPv6 address, hostname addresses. Launch runtime using `DOTNET_DiagnosticPorts` environment variable to setup listener.
```

### Synopsis

```console
dotnet-dsrouter client-client
    [-ipcc|--ipc-client <ipcClient>]
    [-tcpc|--tcp-client <tcpClient>]
    [-rt|--runtime-timeout <timeout>]
    [-v|--verbose <level>]
    [-fp|--forward-port <platform>]
```

### Options

- **`-ipcc, --ipc-client <ipcClient>`**
  The diagnostic tool diagnostics server ipc address (`--diagnostic-port argument`). Router connects diagnostic tool IPC server when establishing a new route between runtime and diagnostic tool.

- **`-tcpc, --tcp-client <tcpClient>`**
  The runtime TCP/IP address using format `[host]:[port]`. Router can connect `127.0.0.1`, `[::1]`, IPv4 address, IPv6 address, hostname addresses. Launch runtime using `DOTNET_DiagnosticPorts` environment variable to setup listener.

- **`-rt, --runtime-timeout <runtimeTimeout>`**
  Automatically shut down router if no runtime connects to it before specified timeout (seconds). If not specified, router won't trigger an automatic shutdown.

- **`-v, --verbose <verbose>`**
  Enable verbose logging (debug|trace)

- **`-fp, --forward-port <forwardPort>`**

>   Enable port forwarding, values Android|iOS for TcpClient and only Android for TcpServer. Make sure to set `ANDROID_SDK_ROOT` before using this option on Android.

Reviewer: ```suggestion
  Enable port forwarding, values `Android` or `iOS` for `TcpClient` and only `Android` for `TcpServer`. Make sure to set `ANDROID_SDK_ROOT` before using this option on Android.
```

## Collect a startup trace using dotnet-trace from a .NET application running on Android

Sometimes it may be useful to collect a trace of an application from its startup. Below illustrates the process of doing so targeting a .NET application running on Android. Since `dotnet-dsrouter` is run using port forwarding, the same scenario will work against applications running in local emulator as well as physical device attached over USB. Make sure to set `ANDROID_SDK_ROOT` before using this option or `dotnet-dsrouter` won't be able to find `adb` needed to setup port forwarding.

- Launch dotnet-dsrouter in server-server mode:

  ```console
  dotnet-dsrouter server-server -ipcs ~/mylocalport -tcps 127.0.0.1:9000 --forward-port Android &
  ```

- Set `DOTNET_DiagnosticPorts` environment variable using `AndroidEnvironment`:

>   Create a file in same folder as `.csproj` using a name like `app.env`, add environment variable into file, `DOTNET_DiagnosticPorts=127.0.0.1:9000,suspend` and include following `ItemGroup` into `.csproj`:

Reviewer: ```suggestion
  Create a file in the same directory as the _.csproj_ using a name like `app.env`, add environment variables into file, `DOTNET_DiagnosticPorts=127.0.0.1:9000,suspend` and include following `ItemGroup` into _.csproj_:
```

  ```xml
  <ItemGroup Condition="'$(AndroidEnableProfiler)'=='true'">
    <AndroidEnvironment Include="app.env" />
  </ItemGroup>
  ```

  It is also possible to set DOTNET_DiagnosticPorts using `adb shell setprop`:

  ```console
  adb shell setprop debug.mono.profile '127.0.0.1:9000,suspend'
  ```

> - Build and launch application using .NET Android SDK and enable tracing by passing `/p:AndroidEnableProfiler=true` to MSBuild. Since application have been configured to suspend on startup, it will connect back to `dotnet-dsrouter` TCP/IP listener running on 127.0.0.1:9000 and wait for diagnostic tooling to connect, before resuming application execution.

Reviewer: ```suggestion
- Build and launch application using .NET Android SDK and enable tracing by passing `/p:AndroidEnableProfiler=true` to MSBuild. Since application have been configured to suspend on startup, it will connect back to `dotnet-dsrouter` TCP/IP listener running on `127.0.0.1:9000` and wait for diagnostic tooling to connect, before resuming application execution.
```

`dotnet-dsrouter` is tied to one running application at any time. If there are needs to trace several different applications at the same time, each application needs to use its own `dotnet-dsrouter` instance, by setting up a unique IPC, TCP/IP address pair in `dotnet-dsrouter` and configure different application instances to connect back to its unique `dotnet-dsrouter` instance.

If `dotnet-dsrouter` is run with `--forward-port` targeting Android and `adb` server or emulator/device gets restarted, all `dotnet-dsrouter` instances needs to be restarted as well to restore port forwarding rules.

When done using `dotnet-dsrouter`, press `<q>` or `<Ctrl+C>` to quit application.

> [!NOTE]
> When running `dotnet-dsrouter` on Windows it will use Named Pipes for its IPC channel. Replace ~/mylocalport with mylocalport in above examples when running on Windows.

> [!NOTE]
> TCP/IP port 9000 is just an example. Any free TCP/IP port can be used.

> [!NOTE]
> Unix Domain Socket ~/mylocalport is just an example. Any free Unix Domain Socket file path can be used.

## Collect a trace using dotnet-trace from a .NET application running on Android

> If there´s no need to collect a trace during application startup, it is possible to launch application in `nosuspend` mode, meaning runtime won't block at startup waiting for diagnostic tooling to connect before resuming execution. Most of the above-described scenario applies to this mode as well, just replace `suspend` with `nosuspend` in `DOTNET_DiagnosticPorts` environment variable to launch application in nosuspend mode.

Reviewer: ```suggestion
If there's no need to collect a trace during application startup, it is possible to launch application in `nosuspend` mode, meaning runtime won't block at startup waiting for diagnostic tooling to connect before resuming execution. Most of the above-described scenario applies to this mode as well, just replace `suspend` with `nosuspend` in `DOTNET_DiagnosticPorts` environment variable to launch application in `nosuspend` mode.
```
